### ISSUE TITLE
parallel runs on single node

### ISSUE THREAD

**Issue by ironv** [optuna/optuna#2456]
I am trying to run Optuna with pyTorch Lightning on an Azure VM with 4 GPUs.  Unfortunately, I do not have sudo access and cannot install mysql or postgres.  The only database which _appears_ to work is sqlite.  I have very simple code to create the study

```
sampler = optuna.samplers.TPESampler(seed=12345)
pruner = optuna.pruners.MedianPruner(n_warmup_steps=10, interval_steps=4)
study = optuna.create_study(storage=f"sqlite:///{clargs['CHKPT_DIR']}/example.db", study_name="test"
                            , load_if_exists=True, sampler=sampler, pruner=pruner, direction="maximize")

obj = Objective(clargs)
study.optimize(obj, n_trials=10, callbacks=[obj.callback])
```
and in the objective function

```
checkpoint_CB = plc.ModelCheckpoint(monitor="val_ap", save_top_k=1, mode="max", dirpath=self.dirpath
                                            , filename=f"trial{trial.number:03}-"+"{epoch:03d}")
earlystopping_CB = plc.early_stopping.EarlyStopping(monitor="val_ap", patience=10, mode="max")
pruning_CB = PyTorchLightningPruningCallback(trial, monitor="val_ap")

print(f"running trial {trial.number}: ", params)
trainer = pl.Trainer(weights_summary=None, gpus=self.gpuid, auto_scale_batch_size="power", auto_lr_find=False
            , deterministic=True, max_epochs=400, logger=False, progress_bar_refresh_rate=0
            , callbacks=[checkpoint_CB,earlystopping_CB,pruning_CB])
trainer.fit(FFNNPL(params))
```

I am kicking off the runs in two terminals and find the GPUs and CPUs are lighting up nicely.  However the parameters chosen are getting repeated.  Please see the output in the post below where I have attached the code.

The trial numbers are incremented perfectly but the parameters are repeated!  Also, with `n_trials = 10`, each terminal does 10 trials for a total of 20.

Questions
[1]  is there something that one can see which I am obviously messing up?
[2] if not, is there is file-based lock/release approach where a study can be loaded in memory parameters saved to it and then the object is saved back to disk so that I do not have to use a database at all.  It may be slower but for some of us that may be the only way.

The other way using `joblib` parallel did not work out (I have added a ticket for that previously).

**Comment by nzw0301** [comment]
HI, @ironv Could you share the codes and commands? I'm afraid that we cannot investigate the issue without full reproducible codes because it takes too much time :(

**Comment by nzw0301** [comment]
@ironv Thank you for sharing your code.

## [1] is there something that one can see which I am obviously messing up?

> The trial numbers are incremented perfectly but the parameters are repeated!

This is because each process uses the same random seed and TPESampler samples hyperparameters independently until the trial id reaches `n_startup_trials`, which default value is `10`.

> Also, with `n_trials = 10`, each terminal does 10 trials for a total of 20.

This is expected behaviour.  (see https://github.com/optuna/optuna/issues/1883).

**Comment by ironv** [comment]
Thank you that certainly helped.  Is there a way of forcing some reproducibility?   I just changed it to `sampler = optuna.samplers.TPESampler(n_startup_trials=5)` and added the `max_trials_callback`.  I could fix a seed per `GPUID` and that would force each run to be different but repeatable (as long as it is on one node).

**Comment by nzw0301** [comment]
I'm glad to hear that. In my understanding, unfortunately, no. We cannot perform the reproducible distributed optimisation. 

[The section in the official FAQ](https://optuna.readthedocs.io/en/stable/faq.html#how-can-i-obtain-reproducible-optimization-results) says 

> First, when optimizing a study in distributed or parallel mode, there is inherent non-determinism. Thus it is very difficult to reproduce the same results in such condition. We recommend executing optimization of a study sequentially if you would like to reproduce the result.
